File: backend/app/schemas/analytics.py
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel


class AnalyticsOverview(BaseModel):
    total_posts: int = 0
    total_published: int = 0
    total_impressions: int = 0
    total_reach: int = 0
    total_likes: int = 0
    total_comments: int = 0
    total_shares: int = 0
    avg_engagement_rate: float = 0.0
    total_followers: int = 0


class PlatformBreakdown(BaseModel):
    platform: str
    platform_username: str
    account_id: str
    posts_count: int = 0
    impressions: int = 0
    reach: int = 0
    likes: int = 0
    comments: int = 0
    shares: int = 0
    followers: int = 0
    engagement_rate: float = 0.0


class PostPerformance(BaseModel):
    post_id: str
    caption: str
    platform: str
    published_at: Optional[datetime] = None
    impressions: int = 0
    reach: int = 0
    likes: int = 0
    comments: int = 0
    shares: int = 0
    engagement_rate: float = 0.0


class DailyMetric(BaseModel):
    date: str
    impressions: int = 0
    reach: int = 0
    likes: int = 0
    comments: int = 0
    engagement_rate: float = 0.0


class AnalyticsDashboard(BaseModel):
    overview: AnalyticsOverview
    platform_breakdown: list[PlatformBreakdown]
    top_posts: list[PostPerformance]
    daily_metrics: list[DailyMetric]


class SnapshotResponse(BaseModel):
    id: str
    social_account_id: str
    metric_type: str
    impressions: int
    reach: int
    likes: int
    comments: int
    shares: int
    saves: int
    clicks: int
    followers_count: int
    engagement_rate: float
    snapshot_date: date
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/api/analytics.ts
import { api } from './client';

export interface AnalyticsOverview {
  total_posts: number;
  total_published: number;
  total_impressions: number;
  total_reach: number;
  total_likes: number;
  total_comments: number;
  total_shares: number;
  avg_engagement_rate: number;
  total_followers: number;
}

export interface PlatformBreakdown {
  platform: string;
  platform_username: string;
  account_id: string;
  posts_count: number;
  impressions: number;
  reach: number;
  likes: number;
  comments: number;
  shares: number;
  followers: number;
  engagement_rate: number;
}

export interface PostPerformance {
  post_id: string;
  caption: string;
  platform: string;
  published_at: string | null;
  impressions: number;
  reach: number;
  likes: number;
  comments: number;
  shares: number;
  engagement_rate: number;
}

export interface DailyMetric {
  date: string;
  impressions: number;
  reach: number;
  likes: number;
  comments: number;
  engagement_rate: number;
}

export interface AnalyticsDashboard {
  overview: AnalyticsOverview;
  platform_breakdown: PlatformBreakdown[];
  top_posts: PostPerformance[];
  daily_metrics: DailyMetric[];
}

export async function getDashboard(days = 30) {
  const resp = await api.get<AnalyticsDashboard>('/analytics/dashboard', {
    params: { days },
  });
  return resp.data;
}

export async function refreshMetrics(accountId: string) {
  const resp = await api.post(`/analytics/refresh/${accountId}`);
  return resp.data;
}
